src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1278,6 +1326,15 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
                     // ISSUE 31315: Determine if/how to expose HTTP2 error codes
                     throw new HttpRequestException(SR.net_http_client_execution_error, e);
                 }
+                else if (e is OperationCanceledException)
+                {
+                    // If the operation has been canceled after the stream was allocated an ID, send a RST_STREAM.
+                    if (http2Stream != null && http2Stream.StreamId != 0)
+                    {
+                        http2Stream.Cancel();
```

Comment:
There's actually a TODO in Http2Stream.Dispose, to add something like this. See: https://github.com/dotnet/corefx/blob/694eeba312046e2097c731d40f32c1cac675aa03/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L358

Note we call Dispose just above in this code. I think Dispose is probably the right place to do this, because we will need to handle this in other exception cases too, e.g. receiving an invalid HEADERS block or something like that.

Reply:
BTW, I was looking at adding some tests for receiving invalid HEADERS blocks. We are definitely not handling this correctly today, and need to do something smarter. Sending the RST_STREAM is somewhat independent of this PR, since we need a way to handle a variety of exceptional cases that will cause RST_STREAM; cancellation is just one of these.

Reply:
I just added #35652 with a bunch of thoughts about this.

Reply:
I'll leave this for a separate PR, if that's okay with you.

Reply:
That's fine.